platform/engineering/request-for-change/2019-09-30-cms-integration.md

```diff
@@ -0,0 +1,237 @@
+site.
+
+Having a Metalsmith build script came from a time when `vets-website`
+had a `content/` directory from which the static pages were
+built. These days, however, all of the content comes from outside
+`vets-website`.
+
+### Content validation
+Some basic validations are performed on the content from the CMS and
+vagov-content repo.
+
+| Validation | Full build | Content-only build | Preview server |
+| ---------- | ---------- | ------------------ | -------------- |
+| Broken link checking | True; will fail the build on the `master` branch only | True; will fail the build | False |
+| Accessibility | True; will fail the build | True; will fail the build | False |
+
+The Metalsmith script performs the check for broken internal
+links. This happens before the accessibility tests in the Jenkins
+build pipelines.
+
+When broken links are discovered during a full Jenkins build, we send
+a Slack notification with a link to the build. If Jenkins was building
+the production (`master`) branch, the build will also fail, to prevent
+broken links in production.
+
+Accessibility tests run the aXe checker on all pages found in the
+sitemap. Whenever the accessibility check fails, the Jenkins build
+fails. This happens after the content build and link checker.
+
+## Design
+I propose rearchitecting the `vets-website` build to separate the
+static content building into a process distinct from the website.
+
+![separate build diagram](images/separated-content-build-proposal.png)
+
+**Note:** Pulling the already-built content _can still_ have two
+different results, much like the current state, but since the content
+it's pulling will have already been validated, the likelihood of a
+failure is drastically reduced.
+
+### Changes
+- The `vets-website` build script will no longer build the content
+  - **Possible exceptions:**
+	- We may need to update some things in the
+	  pre-built content such as references to the new cache-busted
+	  filenames for JS and CSS assets
+	- It should still be responsible for building the pages for React
+	  apps, which would need to have the surrounding markup come from
+	  the templates
+- There will be a new repo for the content build process
+  - This will include scripts and templates
+- There will be a new Jenkins pipeline for building the content
+  - And everything that goes with this
+- There will be a (new?) S3 bucket which holds all the built content
+  - We currently do this with the cache, so this isn't much of a
+    deviation
+- The preview server will live in the new content-building repo
+- Sitemap accessibility tests will be done during the content build,
+  not the `vets-website` build
+- The `vets-website` Jenkins build will have the following changes
+  - There will be no cache fallback
+  - There will be no content caching
+  - There will be no sitemap accessibility testing
+  
+### Serving React apps
+`vets-website` should be responsible for making the pages where the
```

Comment:
Ha! Nope. You didn't miss that at all. That's a great point which I hadn't considered yet. :sweat_smile:

If we had access to `vets-website`'s `manifest.json` files, we could either construct a list of exceptions or create temporary files at the appropriate paths right before doing the link validation. Either way introduces some awkward overhead to get to access the `manifest.json` files. :thinking: I'm not sure yet.

Thank you for pointing this out!

Reply:
Alright, so I took some time to hammer out some possible solutions. I don't know which is the _best_ way, but I think I'm leaning toward using the manifest files in the content build since I prefer to keep the dependencies explicit.

## Problem statement

Currently, these landing pages seem to all be in vagov-content, but as we move away from that, we need to have another solution. We have a Metalsmith plugin which looks at all the manifest.json files for the paths to the React apps and builds a page there for them. When we separate the builds, the content build will be in charge of validating broken links, but won't have access to the React app landing page locations because those are generated from within vets-website.

## Possible solutions

### Keeping the landing pages in Drupal

****Pros****

-   The link checker works with no modification

****Cons****

-   vets-website React apps depend on the CMS having their landing pages
    -   This dependency is implicit
    -   This would also mean that to develop locally, we'd either have to pull down content or still generate temporary landing pages for the applications if none is found...and if the manifest file can't reliably point to where the application lives, how can we tell?
-   The vets-website build would have to update all the landing pages to point to the appropriate assets
    -   How will it know where the landing pages are?
        -   The manifest.json files could point to the wrong landing page if it isn't in charge of where that page actually ends up

### Using the manifest files in the content build

There are at least a couple ways we could validate that the links to React apps aren't broken if we had access to the manifest files.

1.  Add plugins before and after the broken link checker to create and remove dummy files where the React apps would go
2.  Modify the broken link checker to ignore links that are found in an exceptions list that we'd populate from paths found in the manifest files

****Pros****

-   vets-website would be in charge of where the React apps go

****Cons****

-   The content build would be somewhat more complicated
-   The content build would depend on vets-website

### Use blank pages in the content build, but override in the vets-website build

This would look something like the following:

1.  Insert essentially blank pages during the content build
    -   This could either be adding blank pages in Drupal, or as part of the build script itself
2.  The content build validates links to React apps against the blank pages
3.  The vets-website build overrides the blank landing pages
    -   And issues a warning if that page didn't already exist to make sure it gets added into the content build for the content validation
4.  The vets-website production build searches for blank pages
    -   If any are found, alert somehow before deleting

****Pros****

-   The content build wouldn't be strongly tied to vets-website
-   The vets-website build wouldn't be strongly tied to the content

****Cons****

-   The content build would have kind of an implicit dependency on vets-website
    -   It needs to know where the apps are
-   If pages in the content pointed to the blank landing page, but the path was changed from the manifest.json in vets-website, the links will be broken after the blank landing page is removed
    -   Broken links :(
-   The vets-website build would have kind of an implicit dependency on the content build
    -   It needs to know that the dummy landing pages are blank